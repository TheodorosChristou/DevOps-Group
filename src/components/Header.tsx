import { LoginIcon, LogoutIcon } from "@heroicons/react/outline";
import Image from "next/image";
import Link from "next/link";
import { signIn, signOut, useSession } from "next-auth/react";
import { useState } from "react";
import "../i18n";
import { useTranslation } from "react-i18next";
import LanguageSelector from "./LanguageSelector";

export default function Header() {
  const { t, i18n } = useTranslation();
  const { data: session } = useSession();
  const [menuVisible, setMenuVisible] = useState(false);

  const handleSignIn = () => signIn();
  const handleSignOut = () => signOut();

  const toggleMenu = () => {
    setMenuVisible(!menuVisible);
  };

  return (
    <div className="fixed top-0 left-0 right-0 z-40 bg-gray-100">
      <nav className="p-2 items-center w-full flex justify-between">
        <div className="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between">
          <div className="flex align-center">
            <div className="flex space-x-5  ">
            <div className="hidden sm:hidden md:block">

              <Link href="/">
                <Image
                  alt="image"
                  src="/img/mongoose.PNG"
                  width={80}
                  height={80}
                  className="cursor-pointer max-h-sm max-w-lg py-3 mt-0 "
                />
              </Link>
            </div>
            </div>
            <div className="flex space-x-2 ml-1 sm:mt-3 text-black mb-2 font-bold italic text-xl sm:text-2xl sm:py-2 md:py-1 px-0 xs:px-3 sm:px-1 sm:pt-2">
              <a href="/">
                <h1>Mongeese {t("header.map")}</h1>
              </a>
            </div>
          </div>
          <div className="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end">
            <div className="mx-2.5 gap-1 hidden lg:flex">
              <LanguageSelector />
            </div>
            <div className="mx-2.5 gap-1 hidden lg:flex">
              <a
                className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/"
              >
                {t("header.home")}
              </a>
            </div>
            <div className="mx-2.5 gap-1 hidden lg:flex">
              <a
                className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/map"
              >
                {t("header.map")}
              </a>
            </div>
            <div className="mx-2.5 gap-1 hidden lg:flex">
              <a
                className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/legal"
              >
                {t("header.legal")}
              </a>
            </div>
            {session ? (
              <>
                <div className="mx-2.5 gap-1 hidden lg:flex">
                  <a
                    className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                    href="/upload"
                  >
                    {t("header.upload")}
                  </a>
                </div>
                <div className="flex items-center mb-2 mx-2.5">
                  <div>
                    <h1 className="text-5xl cursor-default mb-1">|</h1>
                  </div>
                  <div className="text-ellipsis flex justify-center sm:text-l text-sm mt-2 py-1 px-1 xs:px-3 sm:px-1 cursor-default">
                    <h1>{t("header.welcome")} {session.user.name}</h1>
                  </div>
                  <div
                    className="flex cursor-pointer"
                    onClick={handleSignOut}
                  >
                    <a
                      className="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 px-2 rounded-full mb-1"
                    >
                      <LogoutIcon className="h-6 w-6 mt-1" />
                      {t("header.logout")}
                    </a>
                  </div>
                </div>
              </>
            ) : (
              <div className="flex items-center mb-2 mx-2.5">
                <div>
                  <h1 className="text-5xl cursor-default mb-1">|</h1>
                </div>
                <div
                  className="flex cursor-pointer"
                  onClick={handleSignIn}
                >
                  <a
                    className="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                  >
                    <LoginIcon className="h-6 w-6 mt-1" />
                    {t("header.login")}
                  </a>
                </div>
              </div>
            )}
          </div>
        </div>

        <div className="lg:hidden w-full">
          <div className="flex items-center justify-end">
            <div className="relative inline-block text-left">
              <div>
                <button
                  type="button"
                  className="inline-flex justify-center w-full rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-sm font-medium text-black hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
                  id="menu-button"
                  aria-expanded={menuVisible}
                  aria-haspopup="true"
                  onClick={toggleMenu}
                >
                  Menu
                </button>
              </div>

              {menuVisible && (
                <div
                  className="origin-top-right absolute right-0 mt-2 w-56 rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5"
                  role="menu"
                  aria-orientation="vertical"
                  aria-labelledby="menu-button"
                  tabIndex={-1}
                >
                  <div className="py-1" role="none">
                    <a
                      href="/"
                      className="text-gray-700 block px-4 py-2 text-sm"
                      role="menuitem"
                      tabIndex={-1}
                    >
                      {t("header.home")}
                    </a>
                    <a
                      href="/map"
                      className="text-gray-700 block px-4 py-2 text-sm"
                      role="menuitem"
                      tabIndex={-1}
                    >
                      {t("header.map")}
                    </a>
                    <a
                      href="/legal"
                      className="text-gray-700 block px-4 py-2 text-sm"
                      role="menuitem"
                      tabIndex={-1}
                    >
                      {t("header.legal")}
                    </a>
                    <a
                      href="/documentation"
                      className="text-gray-700 block px-4 py-2 text-sm"
                      role="menuitem"
                      tabIndex={-1}
                    >
                      {t("header.documentation")}
                    </a>
                    {session && (
                      <a
                        href="/upload"
                        className="text-gray-700 block px-4 py-2 text-sm"
                        role="menuitem"
                        tabIndex={-1}
                      >
                        {t("header.upload")}
                      </a>
                    )}
                    <LanguageSelector />
                  </div>
                </div>
              )}
            </div>
          </div>
        </div>
      </nav>
    </div>
  );
}
