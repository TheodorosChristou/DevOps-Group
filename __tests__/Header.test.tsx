import React from 'react'
import {render, screen} from '@testing-library/react'
import Header from '@/components/Header'

jest.mock("next-auth/react", () => {
    const originalModule = jest.requireActual('next-auth/react');
    const mockSession = {
      expires: new Date(Date.now() + 2 * 86400).toISOString(),
      user: { username: "admin" }
    };
    return {
      __esModule: true,
      ...originalModule,
      useSession: jest.fn(() => {
        return {data: mockSession, status: 'authenticated'}  // return type is [] in v3 but changed to {} in v4
      }),
    };
  });
it('Should have a nav bar', () => {
    render(<Header />);
    const navBar = screen.getByTestId('navbar');
    expect(navBar).toHaveClass('p-2')
})

it('Should have an image', () => {
    render(<Header />);
    const image = screen.getByAltText('image');
    expect(image).toBeInTheDocument()
})

it('Should have a heading', () => {
    render (<Header />);
    const heading = screen.getByText('Mongeese Map')
    expect(heading).toBeInTheDocument()
})

 it('Should Show Log Out when has session',
    async () => {
    const {container} = render(<Header/>);
  
    expect(container).toMatchSnapshot()
    expect(screen.getByText("Logout")).toBeInTheDocument();
    })

it('Should have a heading', () => {
    render (<Header />);
    const heading = screen.getByText('Menu')
    expect(heading).toBeInTheDocument()
    })


it('Should Show Menu when has session',
    async () => {
    const {container} = render(<Header/>);
  
    expect(container).toMatchSnapshot()
    expect(screen.getByText("Menu")).toBeInTheDocument();
    })


  
