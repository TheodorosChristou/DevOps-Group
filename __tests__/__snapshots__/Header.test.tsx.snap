// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Show Log Out when has session 1`] = `
<div>
  <div
    class="fixed top-0 left-0 right-0 z-40 bg-gray-100"
  >
    <nav
      class="p-2 items-center w-full flex justify-between"
      data-testid="navbar"
    >
      <div
        class="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between"
      >
        <div
          class="flex align-center"
        >
          <div
            class="flex space-x-5  "
          >
            <div
              class="hidden sm:hidden md:block"
            >
              <a
                href="/"
              >
                <img
                  alt="image"
                  class="cursor-pointer max-h-sm max-w-lg py-3 mt-0 "
                  data-nimg="1"
                  decoding="async"
                  height="80"
                  loading="lazy"
                  src="/_next/image?url=%2Fimg%2Fmongoose.PNG&w=256&q=75"
                  srcset="/_next/image?url=%2Fimg%2Fmongoose.PNG&w=96&q=75 1x, /_next/image?url=%2Fimg%2Fmongoose.PNG&w=256&q=75 2x"
                  style="color: transparent;"
                  width="80"
                />
              </a>
            </div>
          </div>
          <div
            class="flex space-x-2 ml-1 sm:mt-3 text-black mb-2 font-bold italic text-xl sm:text-2xl sm:py-2 md:py-1 px-0 xs:px-3 sm:px-1 sm:pt-2"
          >
            <a
              href="/"
            >
              <h1>
                Mongeese 
                Map
              </h1>
            </a>
          </div>
        </div>
        <div
          class="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end"
        >
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex flex-col items-center"
            >
              <span
                class="font-bold text-black"
              >
                Language
                :
              </span>
              <ul
                class="flex flex-col sm:flex-row space-y-2 sm:space-y-0 sm:space-x-2"
              >
                <li>
                  <button
                    class="text-black hover:text-gray-500 focus:outline-none border-2 border-black px-2 sm:px-3 py-1 rounded-full text-xs sm:text-sm font-bold"
                    type="button"
                  >
                    English
                  </button>
                </li>
                <li>
                  <button
                    class="text-black hover:text-gray-500 focus:outline-none border-2 border-black px-2 sm:px-3 py-1 rounded-full text-xs sm:text-sm font-normal"
                    type="button"
                  >
                    Ελληνικά
                  </button>
                </li>
              </ul>
            </div>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/"
            >
              Home
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/map"
            >
              Map
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/legal"
            >
              Legal
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/documentation"
            >
              Documentation
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/upload"
            >
              Upload
            </a>
          </div>
          <div
            class="flex items-center mb-2 mx-2.5"
          >
            <div>
              <h1
                class="text-5xl cursor-default mb-1"
              >
                |
              </h1>
            </div>
            <div
              class="text-ellipsis flex justify-center sm:text-l text-sm mt-2 py-1 px-1 xs:px-3 sm:px-1 cursor-default"
            >
              <h1>
                Welcome
                 
              </h1>
            </div>
            <div
              class="flex cursor-pointer"
            >
              <a
                class="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 px-2 rounded-full mb-1"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 mt-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Logout
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="lg:hidden w-full"
      >
        <div
          class="flex items-center justify-end"
        >
          <div
            class="relative inline-block text-left"
          >
            <div>
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="inline-flex justify-center w-full rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-sm font-medium text-black hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
                id="menu-button"
                type="button"
              >
                Menu
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;

exports[`Should Show Menu when has session 1`] = `
<div>
  <div
    class="fixed top-0 left-0 right-0 z-40 bg-gray-100"
  >
    <nav
      class="p-2 items-center w-full flex justify-between"
      data-testid="navbar"
    >
      <div
        class="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between"
      >
        <div
          class="flex align-center"
        >
          <div
            class="flex space-x-5  "
          >
            <div
              class="hidden sm:hidden md:block"
            >
              <a
                href="/"
              >
                <img
                  alt="image"
                  class="cursor-pointer max-h-sm max-w-lg py-3 mt-0 "
                  data-nimg="1"
                  decoding="async"
                  height="80"
                  loading="lazy"
                  src="/_next/image?url=%2Fimg%2Fmongoose.PNG&w=256&q=75"
                  srcset="/_next/image?url=%2Fimg%2Fmongoose.PNG&w=96&q=75 1x, /_next/image?url=%2Fimg%2Fmongoose.PNG&w=256&q=75 2x"
                  style="color: transparent;"
                  width="80"
                />
              </a>
            </div>
          </div>
          <div
            class="flex space-x-2 ml-1 sm:mt-3 text-black mb-2 font-bold italic text-xl sm:text-2xl sm:py-2 md:py-1 px-0 xs:px-3 sm:px-1 sm:pt-2"
          >
            <a
              href="/"
            >
              <h1>
                Mongeese 
                Map
              </h1>
            </a>
          </div>
        </div>
        <div
          class="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end"
        >
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex flex-col items-center"
            >
              <span
                class="font-bold text-black"
              >
                Language
                :
              </span>
              <ul
                class="flex flex-col sm:flex-row space-y-2 sm:space-y-0 sm:space-x-2"
              >
                <li>
                  <button
                    class="text-black hover:text-gray-500 focus:outline-none border-2 border-black px-2 sm:px-3 py-1 rounded-full text-xs sm:text-sm font-bold"
                    type="button"
                  >
                    English
                  </button>
                </li>
                <li>
                  <button
                    class="text-black hover:text-gray-500 focus:outline-none border-2 border-black px-2 sm:px-3 py-1 rounded-full text-xs sm:text-sm font-normal"
                    type="button"
                  >
                    Ελληνικά
                  </button>
                </li>
              </ul>
            </div>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/"
            >
              Home
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/map"
            >
              Map
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/legal"
            >
              Legal
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/documentation"
            >
              Documentation
            </a>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <a
              class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              href="/upload"
            >
              Upload
            </a>
          </div>
          <div
            class="flex items-center mb-2 mx-2.5"
          >
            <div>
              <h1
                class="text-5xl cursor-default mb-1"
              >
                |
              </h1>
            </div>
            <div
              class="text-ellipsis flex justify-center sm:text-l text-sm mt-2 py-1 px-1 xs:px-3 sm:px-1 cursor-default"
            >
              <h1>
                Welcome
                 
              </h1>
            </div>
            <div
              class="flex cursor-pointer"
            >
              <a
                class="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 px-2 rounded-full mb-1"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 mt-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                Logout
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="lg:hidden w-full"
      >
        <div
          class="flex items-center justify-end"
        >
          <div
            class="relative inline-block text-left"
          >
            <div>
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="inline-flex justify-center w-full rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-sm font-medium text-black hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
                id="menu-button"
                type="button"
              >
                Menu
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
